pkg/util/k8s/objects.go

```diff
@@ -0,0 +1,191 @@
+package k8s
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	apps_v1 "k8s.io/api/apps/v1"
+	core_v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// CreateObjects will create the provided set of objects. If any object
+// couldn't be created, an error will be returned and created objects will be
+// deleted.
+func CreateObjects(ctx context.Context, cli client.Client, objs ...client.Object) (err error) {
+	// Index offset into objs for objects we managed to create.
+	createdOffset := -1
+
+	defer func() {
+		if err == nil {
+			return
+		}
+		// Delete the subset of objs we managed to create
+		for i := 0; i <= createdOffset; i++ {
+			_ = cli.Delete(context.Background(), objs[i])
+		}
+	}()
+
+	for i, obj := range objs {
+		if err := cli.Create(ctx, obj); err != nil {
+			return fmt.Errorf("failed to create %s: %w", client.ObjectKeyFromObject(obj), err)
+		}
+		createdOffset = i
+	}
+	return nil
+}
+
+// ReadObjects will read the set of objects from r and convert them into
+// client.Object based on the scheme of the provided Kubernetes client.
+//
+// The data of r may be YAML or JSON.
+func ReadObjects(r io.Reader, cli client.Client) ([]client.Object, error) {
+	var (
+		objects []client.Object
+
+		scheme     = cli.Scheme()
+		rawDecoder = yaml.NewYAMLOrJSONDecoder(r, 4096)
+		decoder    = serializer.NewCodecFactory(scheme).UniversalDecoder(scheme.PrioritizedVersionsAllGroups()...)
+	)
+
+NextObject:
```

Comment:
doesn't seem like there's any need for this label

Reply:
<img width="669" alt="image" src="https://user-images.githubusercontent.com/630212/149426641-28ccd39b-6a03-4122-9ef6-037f06beaf9e.png">

`break` in the switch will break out of the switch statement instead of the loop without the label 😞

Reply:
ahh didn't notice that :+1: